src/sdsc/xslt.py

```diff
@@ -0,0 +1,396 @@
+
+    maximumlengths = [24, 35]
+
+    if lengthwarning:
+        try:
+            maximumlengths[0] = int(lengthwarning)
+        except ValueError:
+            printcolor('Sentence length check: Wrong type. Using default.', 'error')
+
+    if lengtherror:
+        try:
+            maximumlengths[1] = int(lengtherror)
+        except ValueError:
+            printcolor('Sentence length check: Wrong type. Using default.', 'error')
+
+    if not content:
+        return []
+
+    messages = []
+    # I get this as a list with one lxml.etree._ElementUnicodeResult, not
+    # as a list with a string.
+    content = str(content[0])
+
+    basefile = basefile[0] if basefile else None
+    contextid = contextid[0] if contextid else None
+
+    # This if/else block should not be necessary (if there is content,
+    # there should always also be pretty content, but that depends on the
+    # XSLT used for checking). It hopefully won't hurt either.
+    contentpretty = str(contentpretty[0]) if contentpretty else content
+
+    sentences = sentencesegmenter(content)
+    # We need to find the current sentence inside of contentpretty by counting tokens.
+    # In content, some tags like <command/> are replaced by
+    # "##@command-<nr tokens>##" so we need to count that as well.
+    sentencestart = 0
+    sentenceend = 0
+
+    for sentence in sentences:
+        words = tokenizer(sentence)
+
+        # Count tag replacements
+        wordcount = 0
+        for token in words:
+            _, _, tagtokens = findtagreplacement(token)
+
+            sentenceend += tagtokens
+            # Tag placeholders count as max. 1 word.
+            wordcount += min(1, tagtokens)
+
+        if wordcount >= maximumlengths[0]:
+            filename = "<file>{0}</file>".format(basefile) if basefile else ""
+            withinid = "<withinid>{0}</withinid>".format(contextid) if contextid else ""
+            messagetype = "error" if wordcount >= maximumlengths[1] else "warning"
+
+            contentpretty = xmlescape(contentpretty)
+            prettytokens = tokenizer(contentpretty)
+            line = linenumber(context)
+            highlightedcontent = highlight(prettytokens, sentencestart, sentenceend - 1)
+            messages.append(etree.XML("""<result type="%s">
+                            <location>%s%s<line>%s</line></location>
+                            <message>Sentence with %s words:
+                            <quote>%s</quote>
+                        </message>
+                        <suggestion>Remove unnecessary words.</suggestion>
+                        <suggestion>Split the sentence.</suggestion>
+                    </result>""" % (messagetype, filename, withinid, str(line),
+                                    str(wordcount), highlightedcontent)))
+
+        sentencestart = sentenceend
+
+    return messages
+
+
+def dupecheck(context, content, contentpretty, contextid, basefile):
+    """TODO@sknorr
```

Comment:
"Takes a paragraph and checks it for duplicated words and phrases of up to three words in length."